File: content/snippets/python/interview_lcci/04/03_list_of_depth.py
from collections import deque


class Node:
    def __init__(self, val: int) -> None:
        self.val = val
        self.left: Node | None = None
        self.right: Node | None = None


def list_of_depth(root: Node | None) -> list[list[int]]:
    if root is None:
        return []
    out: list[list[int]] = []
    q: deque[Node] = deque([root])
    while q:
        level: list[int] = []
        for _ in range(len(q)):
            cur = q.popleft()
            level.append(cur.val)
            if cur.left:
                q.append(cur.left)
            if cur.right:
                q.append(cur.right)
        out.append(level)
    return out


def main() -> None:
    root = Node(1)
    root.left, root.right = Node(2), Node(3)
    root.left.left, root.left.right = Node(4), Node(5)
    root.right.left, root.right.right = Node(6), Node(7)
    print(" | ".join(" ".join(map(str, lvl)) for lvl in list_of_depth(root)))


if __name__ == "__main__":
    main()

File: content/snippets/go/interview_lcci/04/03_list_of_depth.go
package main

import "fmt"

type Node struct {
	Val         int
	Left, Right *Node
}

func listOfDepth(root *Node) [][]int {
	if root == nil {
		return nil
	}
	out := [][]int{}
	q := []*Node{root}
	for len(q) > 0 {
		n := len(q)
		level := []int{}
		for i := 0; i < n; i++ {
			cur := q[0]
			q = q[1:]
			level = append(level, cur.Val)
			if cur.Left != nil {
				q = append(q, cur.Left)
			}
			if cur.Right != nil {
				q = append(q, cur.Right)
			}
		}
		out = append(out, level)
	}
	return out
}

func main() {
	root := &Node{Val: 1}
	root.Left, root.Right = &Node{Val: 2}, &Node{Val: 3}
	root.Left.Left, root.Left.Right = &Node{Val: 4}, &Node{Val: 5}
	root.Right.Left, root.Right.Right = &Node{Val: 6}, &Node{Val: 7}
	levels := listOfDepth(root)
	for i, lvl := range levels {
		if i > 0 {
			fmt.Print(" | ")
		}
		for j, v := range lvl {
			if j > 0 {
				fmt.Print(" ")
			}
			fmt.Print(v)
		}
	}
	fmt.Println()
}
